---
title: "R Assignment - BCB546X"
output: R_notebook
---
Before you grade this assignment, I feel it prudent to begin with the fact I had a hard time with R. I find Unix much for intutive and I have the feeling my set up that I am about to present is much more Unix command line than I would have liked. 

For the first part to this assignment, I used R commands to investigate the properties of the two starting files. The commands I used are below. Looking at the fang file, the dimensions stated that it had 2783 rows and 986 columns. It is a data frame, and of the list type, but the data frame class. It also had no NA values.

The snp file, has dimensions of 1017 rows, and 3 columns after processing. It is a list type, data frame class, and has no NA values.

```{r}
dim(fang_data)
dim(snp_position)

typeof(fang_data)
typeof(snp_position)

class(fang_data)
class(snp_position)

sum(is.na(fang_data))
sum(is.na(snp_position))

```

For the second part of this assignment, I spent four days trying to get this code to work. I discovered I have no clue how R for loops work and I really wish I did. Either way, I ended up writing out all the commands for each chromosome because I still could not get anything other than a basic for loop to work. 

```{r}
fang_data <- read.table("fang_et_al_genotypes.txt")
snp_position <- read.table("snp_position.txt", header = TRUE, fill = TRUE)

#Maize Commands
maize_genotypes <- fang_data[fang_data$V3 %in% c("ZMMIL", "ZMMLR", "ZMMMR", "Group"),]

transposed_maize <- t(maize_genotypes)
#Transposed data was as a matrix and not a data frame.

transposed_maize <- as.data.frame(transposed_maize)
#This command turned it back to a data frame.

snp_position <- snp_position[,-2]
snp_position <- snp_position[,1:3]

merged_maize <- merge(snp_position, transposed_maize, by = 1, all.x = F, all.y = F)

list_of_df <- list()

for (i in 1:10) { 
   list_of_df[[i]] <- merged_maize[merged_maize$Chromosome == i,]
   assign(paste("maize_chr_",i,sep=""), list_of_df[[i]])

   }

#I struggled for a long time trying to figure out the rest of the for loop. 
#I didn't succeed and just typed it all out.
#It's not pretty, but it gets the job done.
#Accessing the variables in a group was my main issue.

maize_chr_1$Position <- as.numeric(as.character(maize_chr_1$Position))
maize_chr_1_ascending <- maize_chr_1[order(maize_chr_1$Position),]
maize_chr_1_descending <- maize_chr_1[order(-maize_chr_1$Position),]
maize_chr_1_descending[] <- lapply(maize_chr_1_descending, gsub, pattern = "?", replacement = "-", fixed = TRUE)

write.table(maize_chr_1_ascending,
  file="maize_chr_1_ascending.csv",
  sep=",", quote=FALSE, row.names=FALSE)
write.table(maize_chr_1_descending,
            file="maize_chr_1_descending.csv",
            sep=",", quote=FALSE, row.names=FALSE)
```

This code is not elegant, nor is it pretty, but it works to get the files that we need. I came away from this section feeling like I had a very good idea of what good uses for R were, and it was not this type of assigment. I used the same commands for the teosinte data, except I replaced "maize" with "teosinte" in the file names.

In addition, I repeated all the code for other other chromosomes. 

For the chromosome graphing section, I used the following code:

```{r}
maize_chr_1 <- maize_chr_1[,-(1:3)]
teosinte_chr_1 <- teosinte_chr_1[,-(1:3)]
snp_count_chr_1 <- (nrow(maize_chr_1)*ncol(maize_chr_1)) + (nrow(teosinte_chr_1)*ncol(teosinte_chr_1))

maize_chr_2 <- maize_chr_2[,-(1:3)]
teosinte_chr_2 <- teosinte_chr_2[,-(1:3)]
snp_count_chr_2 <- (nrow(maize_chr_2)*ncol(maize_chr_2)) + (nrow(teosinte_chr_2)*ncol(teosinte_chr_2))

maize_chr_3 <- maize_chr_3[,-(1:3)]
teosinte_chr_3 <- teosinte_chr_3[,-(1:3)]
snp_count_chr_3 <- (nrow(maize_chr_3)*ncol(maize_chr_3)) + (nrow(teosinte_chr_3)*ncol(teosinte_chr_3))

maize_chr_4 <- maize_chr_4[,-(1:3)]
teosinte_chr_4 <- teosinte_chr_4[,-(1:3)]
snp_count_chr_4 <- (nrow(maize_chr_4)*ncol(maize_chr_4)) + (nrow(teosinte_chr_4)*ncol(teosinte_chr_4))

maize_chr_5 <- maize_chr_5[,-(1:3)]
teosinte_chr_5 <- teosinte_chr_5[,-(1:3)]
snp_count_chr_5 <- (nrow(maize_chr_5)*ncol(maize_chr_5)) + (nrow(teosinte_chr_5)*ncol(teosinte_chr_5))

maize_chr_6 <- maize_chr_6[,-(1:3)]
teosinte_chr_6 <- teosinte_chr_6[,-(1:3)]
snp_count_chr_6 <- (nrow(maize_chr_6)*ncol(maize_chr_6)) + (nrow(teosinte_chr_6)*ncol(teosinte_chr_6))

maize_chr_7 <- maize_chr_7[,-(1:3)]
teosinte_chr_7 <- teosinte_chr_7[,-(1:3)]
snp_count_chr_7 <- (nrow(maize_chr_7)*ncol(maize_chr_7)) + (nrow(teosinte_chr_7)*ncol(teosinte_chr_7))

maize_chr_8 <- maize_chr_8[,-(1:3)]
teosinte_chr_8 <- teosinte_chr_8[,-(1:3)]
snp_count_chr_8 <- (nrow(maize_chr_8)*ncol(maize_chr_8)) + (nrow(teosinte_chr_8)*ncol(teosinte_chr_8))

maize_chr_9 <- maize_chr_9[,-(1:3)]
teosinte_chr_9 <- teosinte_chr_9[,-(1:3)]
snp_count_chr_9 <- (nrow(maize_chr_9)*ncol(maize_chr_9)) + (nrow(teosinte_chr_9)*ncol(teosinte_chr_9))

maize_chr_10 <- maize_chr_10[,-(1:3)]
teosinte_chr_10 <- teosinte_chr_10[,-(1:3)]
snp_count_chr_10 <- (nrow(maize_chr_10)*ncol(maize_chr_10)) + (nrow(teosinte_chr_10)*ncol(teosinte_chr_10))

total_snp_count <- append(snp_count_chr_1, snp_count_chr_2)
total_snp_count <- append(total_snp_count, snp_count_chr_3)
total_snp_count <- append(total_snp_count, snp_count_chr_4)
total_snp_count <- append(total_snp_count, snp_count_chr_5)
total_snp_count <- append(total_snp_count, snp_count_chr_6)
total_snp_count <- append(total_snp_count, snp_count_chr_7)
total_snp_count <- append(total_snp_count, snp_count_chr_8)
total_snp_count <- append(total_snp_count, snp_count_chr_9)
total_snp_count <- append(total_snp_count, snp_count_chr_10)

total_snp_count <- as.data.frame(total_snp_count)
total_snp_count$chromosome <- c(1:10)
ggplot(total_snp_count, aes(chromosome, total_snp_count, fill=as.factor(chromosome))) + geom_bar(stat="identity")

```

For it, I made a file of total snp counts using the codes from each chromosome file I made. Then, I added these all together in one list, and added another column with chromosome names. From there, I converted the list to a data frame and graphed using gglot2.

I saved everything to the Git repository, including the notebook, the original README file, and my graph.