# Unix_Assignment

In this assignment, we were responsible for inspecting, describing, and manipulating SNP data for downstream analysis. In all, 40 files were created with data pertaining to maize and teosinte SNP data. Each file represents the SNP on a particular chromosome and particular order of the SNP position data. If done correctly, these files should be able to be be easily used for further analysis.

## Inspection and Describing

The inspection part of this project involved using UNIX commands to investigate different fundamental aspects of the original files before use. The two files are fang_et_al_genotypes.txt and snp_position.txt. These files are not included in the repository, as they were given by the instructor.

### fang_et_al_genotypes.txt

I began this inspection using the wc command. It showed that this file had 2783 lines, 2744038 words, and was made up of 11051939 bytes of information. The du -h command additionally showed that this file was 11 megabytes in total, definitely one of the larger files I have every worked with. Continuing my inspection, I used:

$ tail -n +6 fang_et_al_genotypes.txt | awk -F "\t" '{print NF; exit}' 

to show that the fang file has 986 columns in addition to the 2783 lines. The tail made it so this value did not include the header. I also ran the same command without the tail and the pipe, and got the same value. This showed that the header was not affected the number of columns. The column command showed that the first three column names are Sample_ID, JG_OTU, and Group. It does not appear that the column position needs to be changed. 

### snp_position.txt

For the snp_position.txt file, I ran all the same commands used above on fang_et_al_genotypes.txt. As such, I will not bother to repeat them here. For the snp_position.txt, there are 984 line, 13198 words, and is comprised of 82763 bytes. In addition, it is approximately 84 kilobytes. This is expected, considering the wc generated byte count. The column counting awk command showed that in contrast to the above file, snp_position has 15 columns. As above, the command was also done without the tail command and pipe. This showed the column number to be the same with or without the header. The column command showed that the first three columns for snp_position are SNP_ID, cdv_marker_id, and Chromosome. The fourth column is named Position. This indicates that the second column of snp_position will need to be removed in order to get the proper final file format after joining.

## Data Manipulation

This section will detail the workflow and commands that I used to edit the files and create the final 40 data files asked for in this project. As a note, I will only detail the commands used for one organism, maize. The exact same commands were used for teosinte, with only file names changing per command. 

To begin, I saw that we needed to get rid of the second column in the snp_position file. To do this, I used the following unix command:

$ cut -f 1,3,4 snp_position.txt > cut_snp_position.txt

This command was supposed to take the first column, third column, and fourth column of snp_position and put this output to a new file called cut_snp_position.txt. I checked with the column command and it appeared to do so. The other columns in snp_position I did not think were necessary for the project and therefore did not include them. For joining and sorting, and only the first column, the third column, However, if they needed to be added for downstream analysis, the command could be changed to the following:

$ cut -f 1,3-15 snp_position.txt > cut_snp_position.txt

With the snp_position file taken care of, I turned to the fang_et_al_genotypes file. In order for proper analysis, maize and teosinte samples had to be separated. I used a grep command to separate them based on the group ID provided in the project description. The command was as follows:

$ grep -E "(Group|ZMMIL|ZMMLR|ZMMMR)" fang_et_al_genotypes.txt > maize_genotypes.txt

This pulled the header, and all lines relating to maize into one file. Following this, I used the transpose.awk file to transposed the genotype file for proper joining. After transposing the file, the genotype file and the cut_snp file were joined and sorted in one command via a pipe. The commands were used as follows:

$ awk -f transpose.awk maize_genotypes.txt > transposed_maize_genotypes.txt
$ join -t $'\t' -1 1 -2 1 cut_snp_position.txt transposed_maize_genotypes.txt | sort -k2,2n > joined_maize_snps.txt

After the files were joined and sorted, it was necessary to separate all the SNPs on the same chromosome into individual files. The way I chose to do this was by using an awk command individually for each chromosome. I was trying to figure out a way to separate them all using only one command once, but I couldnâ€™t. Any feedback you have about how to do this would be great. For the first chromosome, I had to specify that I was looking for 1 and not 10, as my command without the additional && segment put 1 and 10 in the same file. In addition, I piped the output of the awk command to another sort command so that the resulting file was only one chromosome and was sorted from smallest to larger position value. The command was:

$ awk '$2 ~ /1/ && $2 != 10' sorted_joined_maize_.txt | sort -k3,3n > increasing_maize_chr1.txt

With the above command, it was unnecessary to change any of the symbols for missing data, as they were question marks and did not need to be changed. However, for the file of the chromosome data sorted from largest to smallest position value, the missing data figure had to be replaced with a -. To do this, and to sort the file in decreasing order, I used the following command:

$ sed 's/?/-/g' increasing_maize_chr1.txt | sort -k3,3nr > decreasing_maize_chr1.txt

These sets of commands were repeated for all the other chromosomal data excluding the && segment in the awk command for chromosomes 2 through 9, and including it in 10 with the difference between that $2 != 1 instead of cannot equal 10. 

## Git Repository

To turn in this assignment, I created a Git repository online labelled Unix_Assignment, created a README.md file to initialize, and then pull the folder to my local computer. I had been working on these files locally before putting in the Git repository, something I should not have been doing, but I forgotten I was working in my play folder system when I figured out how to do the assignment. I moved the files to the local repository file, added them using the git add * command, committed them, and then pushed the final files to the Git repository.